src/engine/threaded_engine.cc

```diff
@@ -391,18 +394,29 @@ void ThreadedEngine::WaitForAll() {
   finished_cv_.wait(lock, [this]() {
       return pending_.load() == 0 || kill_.load();
     });
+
+  if (global_exc_waitall_ptr) {
+    std::rethrow_exception(global_exc_waitall_ptr);
+  }
 }
 
 inline void ThreadedEngine::OnComplete(ThreadedOpr* threaded_opr) {
   bool is_temporary_opr = threaded_opr->temporary;
   // Mark complete for read variables
   for (auto&& i : threaded_opr->const_vars) {
+    if (i->var_ex) {
+        global_exc_waitall_ptr = i->var_ex;
+        threaded_opr->opr_ex = i->var_ex;
+    }
     i->CompleteReadDependency([this](OprBlock* opr) {
         this->PushToExecute(opr, false);
       });
   }
   // Mark complete for write variables.
   for (auto&& i : threaded_opr->mutable_vars) {
+    if (threaded_opr->opr_ex) {
```

Comment:
Instead of each operator holding its exception and having to iterate through them afterwards, would a more generalized approach be simpler where a pointer to a shared exception-container object is in each threaded_opr? It could be a queue or list or just one. I'd imagine we only care about the first exception, right?

Reply:
^^ Reasoning is that if there is a highly-generalized approach, then it can be easily reused, for example by a templated call wrapper or something like that.

Reply:
@cjolivier01 ThreadedOpr and ThreadedVar has std::exception_ptr which is a shared ownership smart pointer as its member. Isn't this the same as what you are suggesting ?